gcsmigrate/shared.py

```diff
@@ -0,0 +1,354 @@
+        )
+
+    def to_json(self):
+        """Serialize as JSON."""
+        return bytes(
+            json.dumps({"length": self.length, "fid": self.fid, "dkey": self.dkey}),
+            "utf-8",
+        )
+
+    @classmethod
+    def from_json(cls, json_str):
+        """Create MogileFile from JSON string."""
+        return MogileFile(**json.loads(json_str))
+
+    def verify(self, fid, md5, sha1, bucket, bucket_map, gcs_client):
+        """Verify (with assert) this mogile file against asserted values."""
+        print(f"Verifying {self.fid}")
+        assert self.fid == fid, f"fid {self.fid} not migrated"
+        new_bucket = bucket_map[self.mogile_bucket]
+        assert new_bucket == bucket
+        remote_md5 = self.contentrepo_exists_in_bucket(gcs_client, new_bucket)
+        assert remote_md5 == md5, f"{self.fid} has wrong MD5 sum"
+        assert self.sha1sum == sha1, f"{self.fid} has wrong SHA1 sum"
+
+
+def get_mogile_files_from_database(
+    database_url, limit=None, fids=None, excluded_fids=set()
+):
+    """Return a generator for all mogile files in the database."""
+    config = dj_database_url.parse(database_url)
+    connection = pymysql.connect(
+        host=config["HOST"],
+        user=config["USER"],
+        password=config["PASSWORD"],
+        db=config["NAME"],
+        cursorclass=pymysql.cursors.SSCursor,
+    )
+
+    try:
+        cursor = connection.cursor()
+        if limit is not None:
+            sql = f"SELECT * FROM file LIMIT {limit}"
+        elif fids is not None:
+            fids_in = ", ".join(fids)
+            sql = f"SELECT * FROM file WHERE fid IN ({fids_in})"
+        else:
+            sql = "SELECT * FROM file"
+        cursor.execute(sql)
+        row = cursor.fetchone()
+        while row:
+            if row[0] not in excluded_fids:
+                yield MogileFile.parse_row(row)
+            row = cursor.fetchone()
+    finally:
+        connection.close()
+
+
+def make_generator_from_args(args):
+    """Make a mogile_file generator from args.
+
+    Args are either a list of fids to process or a single file that
+    contains a list of fids to exclude.
+    """
+    fids = None
+    excluded_fids = set()
+    if len(args) > 0:
+        if args[0].isdigit():
+            fids = args
+        else:
+            with open(args[0]) as f:
+                for line in f:
+                    excluded_fids.add(int(line))
+            print(f"Excluding {len(excluded_fids)} fids.")
+    return get_mogile_files_from_database(
+        os.environ["MOGILE_DATABASE_URL"], fids=fids, excluded_fids=excluded_fids
+    )
+
+
+def chunked(iterable, size):
+    """Group into chunks of size."""
+    result = []
+    for item in iterable:
+        result.append(item)
+        if len(result) == size:
+            yield result
+            result = []
+    if len(result) > 0:
+        yield result
+
+
+def future_waiter(iterable, max_futures):
```

Comment:
did you try using the standard lib function here? https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait Would that work? (would it have the right semantics?)

Reply:
also, what happens if any of these futures fail? Will they throw an exception?

Reply:
I thought I replied to this, but my reply seems to have been lost. I don't think `wait` would work, but I thought `as_completed` would. Unfortunately, it turns out that `as_completed` uses some hidden method or something of the future, and while google's future (https://googleapis.dev/python/pubsub/latest/subscriber/api/futures.html) seems to support the public interface of python's future, it doesn't support some of the `_` methods. So using `as_completed` gave me an error.

Good point on failure. Let me fix that.

Reply:
2d8f520

Reply:
Looks great! Thanks!